### PR TITLE
test: adding more tests to a star algorithm

### PR DESCRIPTION
**Pull request by kibolho**
### Describe your change:
Adding more tests to a star algorithm
Fixes #9943 

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

File: graphs/a_star.py [hunk 1]
```diff
@@ -16,6 +16,30 @@ def search(
     cost: int,
     heuristic: list[list[int]],
 ) -> tuple[list[list[int]], list[list[int]]]:
+    """
+    Search for a path on a grid avoiding obstacles.
+    >>> grid = [[0, 1, 0, 0, 0, 0], \
+                [0, 1, 0, 0, 0, 0], \
+                [0, 1, 0, 0, 0, 0], \
+                [0, 1, 0, 0, 1, 0], \
+                [0, 0, 0, 0, 1, 0]]
```
Comment (tianyizheng02) [inline graphs/a_star.py:25]:
```suggestion
    >>> grid = [[0, 1, 0, 0, 0, 0],
    ...         [0, 1, 0, 0, 0, 0],
    ...         [0, 1, 0, 0, 0, 0],
    ...         [0, 1, 0, 0, 1, 0],
    ...         [0, 0, 0, 0, 1, 0]]
```

File: graphs/a_star.py [hunk 2]
```diff
@@ -16,6 +16,30 @@ def search(
...
+    >>> init = [0, 0]
+    >>> goal = [len(grid)-1, len(grid[0])-1]
```
Comment (tianyizheng02) [inline graphs/a_star.py:27]:
```suggestion
    >>> goal = [len(grid) - 1, len(grid[0]) - 1]
```

File: graphs/a_star.py [hunk 3]
```diff
@@ -16,6 +16,30 @@ def search(
...
+    >>> cost = 1
+    >>> heuristic = [[0 for row in range(len(grid[0]))] for col in range(len(grid))]
```
Comment (tianyizheng02) [inline graphs/a_star.py:29]:
```suggestion
    >>> heuristic = [[0] * len(grid[0]) for _ in range(len(grid))]
```

File: graphs/a_star.py [hunk 4]
```diff
@@ -16,6 +16,30 @@ def search(
...
+    >>> for i in range(len(grid)):
+    ...     for j in range(len(grid[0])):
+    ...         heuristic[i][j] = abs(i - goal[0]) + abs(j - goal[1])
+    ...         if grid[i][j] == 1:
+    ...             heuristic[i][j] = 99
+    >>> path, action = search(grid, init, goal, cost, heuristic)
+    >>> print(path)
+    [[0, 0], [1, 0], [2, 0], [3, 0], [4, 0], [4, 1], [4, 2],\
+ [4, 3], [3, 3], [2, 3], [2, 4], [2, 5], [3, 5], [4, 5]]
```
Comment (tianyizheng02) [inline graphs/a_star.py:38]:
```suggestion
    >>> path
    [[0, 0], [1, 0], [2, 0], [3, 0], [4, 0], [4, 1], [4, 2],
    ... [4, 3], [3, 3], [2, 3], [2, 4], [2, 5], [3, 5], [4, 5]]
```

File: graphs/a_star.py [hunk 5]
```diff
@@ -16,6 +16,30 @@ def search(
...
+    >>> print(action)
+    [[0, 0, 0, 0, 0, 0], [2, 0, 0, 0, 0, 0], [2, 0, 0, 0, 3, 3],\
+ [2, 0, 0, 0, 0, 2], [2, 3, 3, 3, 0, 2]]
```
Comment (tianyizheng02) [inline graphs/a_star.py:41]:
```suggestion
    >>> action
    [[0, 0, 0, 0, 0, 0], [2, 0, 0, 0, 0, 0], [2, 0, 0, 0, 3, 3],
    ... [2, 0, 0, 0, 0, 2], [2, 3, 3, 3, 0, 2]]
```

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]
Minor formatting tweaks

File: graphs/a_star.py (hunk 4 above)
Comment (kibolho) [reply graphs/a_star.py]: Check this in the CI: https://github.com/TheAlgorithms/Python/actions/runs/6517757487/job/17702529065?pr=10397

File: graphs/a_star.py (hunk 5 above)
Comment (kibolho) [reply graphs/a_star.py]: > Check this in the CI: https://github.com/TheAlgorithms/Python/actions/runs/6517757487/job/17702529065?pr=10397

File: graphs/a_star.py [hunk 6]
```diff
@@ -16,6 +16,31 @@ def search(
...
+    >>> heuristic = [[0 for row in range(len(grid[0]))] for col in range(len(grid))]
+    >>> for i in range(len(grid)):
+    ...     for j in range(len(grid[0])):
+    ...         heuristic[i][j] = abs(i - goal[0]) + abs(j - goal[1])
+    ...         if grid[i][j] == 1:
+    ...             heuristic[i][j] = 99
+    >>> path, action = search(grid, init, goal, cost, heuristic)
+    >>> path
+    [[0, 0], [1, 0], [2, 0], [3, 0], [4, 0], [4, 1], [4, 2],\
+ [4, 3], [3, 3], [2, 3], [2, 4], [2, 5], [3, 5], [4, 5]]
```
Comment (tianyizheng02) [inline graphs/a_star.py:39]:
```suggestion
    >>> path
    [[0, 0], [1, 0], [2, 0], [3, 0], [4, 0], [4, 1],  # doctest: +NORMALIZE_WHITESPACE
    [4, 2], [4, 3], [3, 3], [2, 3], [2, 4], [2, 5], [3, 5], [4, 5]]
```

File: graphs/a_star.py [hunk 7]
```diff
@@ -16,6 +16,31 @@ def search(
...
+    >>> action
+    [[0, 0, 0, 0, 0, 0], [2, 0, 0, 0, 0, 0], [2, 0, 0, 0, 3, 3],\
+ [2, 0, 0, 0, 0, 2], [2, 3, 3, 3, 0, 2]]
```
Comment (tianyizheng02) [inline graphs/a_star.py:42]:
```suggestion
    >>> action
    [[0, 0, 0, 0, 0, 0], [2, 0, 0, 0, 0, 0],  # doctest: +NORMALIZE_WHITESPACE
    [2, 0, 0, 0, 3, 3], [2, 0, 0, 0, 0, 2], [2, 3, 3, 3, 0, 2]]
```

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

File: graphs/a_star.py (hunk 5 above)
Comment (kibolho) [reply graphs/a_star.py]: Nice @tianyizheng02 Thanks!

File: graphs/a_star.py (hunk 4 above)
Comment (kibolho) [reply graphs/a_star.py]: Nice @tianyizheng02 Thanks!

**Review by tianyizheng02 — APPROVED** [review]

**Comment by tianyizheng02** [comment]
Thanks for your contribution and for your fixes!

**Pull request merged**